test/mlir/test/gc/Transforms/buffer-merge.mlir

```diff
@@ -176,4 +176,185 @@ func.func @different_mem_space() {
   "test.source"(%defaultv2)  : (memref<8x64xf32>) -> ()
   "test.source"(%defaultv)  : (memref<8x64xf32>) -> ()
   return
-}
\ No newline at end of file
+}
+
+func.func @nested_forall(%arg0: memref<2xf32>) {
+    %c4 = arith.constant 4 : index
+    %c16 = arith.constant 16 : index
+    %alloc = memref.alloc() : memref<2xf32>
+    memref.copy %arg0, %alloc : memref<2xf32> to memref<2xf32>
+    scf.forall (%i) in (%c16) {
+      %alloc_0 = memref.alloc() : memref<4xf32>
+      %alloc_1 = memref.alloc() : memref<8xf32> // hoist to outermost
+      scf.forall (%j) in (%c4) {
+        "test.source"(%alloc) : (memref<2xf32>) -> ()
+        "test.source"(%alloc_1) : (memref<8xf32>) -> ()
+        %alloc_2 = memref.alloc() : memref<16xf32>
+        "test.source"(%alloc_2) : (memref<16xf32>) -> ()
+      }
+      "test.source"(%alloc) : (memref<2xf32>) -> ()
+      "test.source"(%alloc_0) : (memref<4xf32>) -> ()
+    }
+    return
+}
+// CHECK-LABEL: func @nested_forall
+//       CHECK: %[[ALLOC0:.*]] = memref.alloc() {alignment = 64 : i64} : memref<576xi8>
+//       CHECK: %[[VIEW0:.*]] = memref.view %[[ALLOC0]][%{{.*}}][] : memref<576xi8> to memref<2xf32>
+//  CHECK-NEXT: memref.copy %arg0, %[[VIEW0]]
+//  CHECK-NEXT: scf.forall
+//       CHECK:   %[[ALLOC1:.*]] = memref.alloc() {alignment = 64 : i64} : memref<64xi8>
+//       CHECK:   %[[VIEW1:.*]] = memref.view %[[ALLOC1]][%{{.*}}][] : memref<64xi8> to memref<4xf32>
+//       CHECK:   %[[VIEW2:.*]] = memref.view %[[ALLOC0]][%{{.*}}][] : memref<576xi8> to memref<8xf32>
+//  CHECK-NEXT:   scf.forall
+//  CHECK-NEXT:     %[[ALLOC2:.*]] = memref.alloc() {alignment = 64 : i64} : memref<64xi8>
+//       CHECK:     %[[VIEW3:.*]] = memref.view %[[ALLOC2]][%{{.*}}][] : memref<64xi8> to memref<16xf32>
+
+func.func @mixed_forall_and_for(
+    %lb: index,
+    %ub: index,
+    %step: index,
+    %arg0: memref<2xf32>) {
+    %c4 = arith.constant 4 : index
+    %c16 = arith.constant 16 : index
+    %0 = memref.alloc() : memref<2xf32>
+    scf.forall (%i) in (%c16) {
+      %alloc = memref.alloc() : memref<4xf32> //hoist to outermost
+      scf.forall (%j) in (%c4) {
+        %alloc_0 = memref.alloc() : memref<8xf32>
+        %1 = scf.for %k = %lb to %ub step %step
+          iter_args(%iterBuf = %arg0) -> (memref<2xf32>) {
```

Comment:
We generally don't set memref in `iter_args`. We will directly write to `arg0` after bufferization. `iter_args` is usually for tensors.